File: Battleships/Leaderboard.cpp
#include "Leaderboard.hpp"
#include "Color.hpp"
#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>

Leaderboard::Leaderboard() {
    loadFromFile();
}

void Leaderboard::loadFromFile() {
    std::ifstream file(filename);
    if (!file.is_open()) {
        return;
    }
    std::string name;
    int wins;
    while (file >> name >> wins) {
        records[name] = wins;
    }
    file.close();
}

void Leaderboard::saveToFile() {
    std::ofstream file(filename);
    if (!file.is_open()) {
        std::cout << "Îøèáêà ñîõðàíåíèÿ òàáëèöû ëèäåðîâ!" << std::endl;
        return;
    }
    for (const auto& record : records) {
        file << record.first << " " << record.second << std::endl;
    }
    file.close();
}

void Leaderboard::addWin(const std::string& playerName) {
    records[playerName]++;
    saveToFile();
    Color::setColor(Color::GREEN);
    std::cout << "Ïîáåäà èãðîêà " << playerName << " ñîõðàíåíà â òàáëèöå ëèäåðîâ!" << std::endl;
    Color::resetColor();
    display();
}


void Leaderboard::display() const {
    std::cout << "\n";
    Color::setColor(Color::YELLOW);
    std::cout << "=== ÒÀÁËÈÖÀ ËÈÄÅÐÎÂ ===" << std::endl;
    Color::resetColor();
    if (records.empty()) {
        std::cout << "Ïîêà íåò çàïèñåé." << std::endl;
        return;
    }
    std::vector<std::pair<std::string, int>> sortedRecords;
    for (const auto& record : records) {
        sortedRecords.push_back(record);
    }
    for (size_t i = 0; i < sortedRecords.size() - 1; ++i) {
        for (size_t j = i + 1; j < sortedRecords.size(); ++j) {
            if (sortedRecords[i].second < sortedRecords[j].second) {
                std::swap(sortedRecords[i], sortedRecords[j]);
            }
        }
    }
    std::cout << "------------------------------" << std::endl;
    std::cout << "Èãðîê\t\tÏîáåäû" << std::endl;
    std::cout << "------------------------------" << std::endl;
    for (size_t i = 0; i < sortedRecords.size(); ++i) {
        const auto& record = sortedRecords[i];
        if (i == 0) {
            Color::setColor(Color::YELLOW);  
        }
        else if (i == 1) {
            Color::setColor(Color::GRAY);    
        }
        else if (i == 2) {
            Color::setColor(Color::RED);     
        }
        else {
            Color::setColor(Color::WHITE);   
        }
        std::cout << record.first;
        if (record.first.length() < 8) {
            std::cout << "\t\t";
        }
        else {
            std::cout << "\t";
        }
        std::cout << record.second << std::endl;
        Color::resetColor();
    }
    std::cout << "------------------------------" << std::endl;
}

File: SeaBattleCSharp/Leaderboard.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace SeaBattleCSharp
{
    public class Leaderboard
    {
        private Dictionary<string, int> records;
        private const string filename = "leaderboard.txt";

        public Leaderboard()
        {
            records = new Dictionary<string, int>();
            LoadFromFile();
        }

        private void LoadFromFile()
        {
            if (!File.Exists(filename))
                return;

            try
            {
                string[] lines = File.ReadAllLines(filename);
                foreach (string line in lines)
                {
                    string[] parts = line.Split(' ');
                    if (parts.Length >= 2 && int.TryParse(parts[1], out int wins))
                    {
                        records[parts[0]] = wins;
                    }
                }
            }
            catch
            {
                Console.WriteLine("Ошибка загрузки таблицы лидеров!");
            }
        }

        private void SaveToFile()
        {
            try
            {
                using (StreamWriter file = new StreamWriter(filename))
                {
                    foreach (var record in records)
                    {
                        file.WriteLine($"{record.Key} {record.Value}");
                    }
                }
            }
            catch
            {
                Console.WriteLine("Ошибка сохранения таблицы лидеров!");
            }
        }

        public void AddWin(string playerName)
        {
            if (records.ContainsKey(playerName))
                records[playerName]++;
            else
                records[playerName] = 1;

            SaveToFile();
            Color.SetColor(Color.GREEN);
            Console.WriteLine($"Победа игрока {playerName} сохранена в таблице лидеров!");
            Color.ResetColor();
            Display();
        }

        public void Display()
        {
            Console.WriteLine("\n");
            Color.SetColor(Color.YELLOW);
            Console.WriteLine("=== ТАБЛИЦА ЛИДЕРОВ ===");
            Color.ResetColor();

            if (records.Count == 0)
            {
                Console.WriteLine("Пока нет записей.");
                return;
            }

            var sortedRecords = records.ToList();
            sortedRecords.Sort((a, b) => b.Value.CompareTo(a.Value));

            Console.WriteLine("------------------------------");
            Console.WriteLine("Игрок\t\tПобеды");
            Console.WriteLine("------------------------------");

            for (int i = 0; i < sortedRecords.Count; i++)
            {
                var record = sortedRecords[i];

                if (i == 0)
                    Color.SetColor(Color.YELLOW);
                else if (i == 1)
                    Color.SetColor(Color.GRAY);
                else if (i == 2)
                    Color.SetColor(Color.RED);
                else
                    Color.SetColor(Color.WHITE);

                Console.Write(record.Key);
                if (record.Key.Length < 8)
                    Console.Write("\t\t");
                else
                    Console.Write("\t");

                Console.WriteLine(record.Value);
                Color.ResetColor();
            }

            Console.WriteLine("------------------------------");
        }
    }
}